docs/src/proposals/comprehensive-compute-fees-review.md

```diff
@@ -0,0 +1,164 @@
+
+### Fees for precompiled program failures
+
+https://github.com/solana-labs/solana/issues/20481
+
+### Rate governing
+
+Current rate governing needs to be re-assessed.  Fees are being rate
+governed down to their minimums because the number of signatures in each slot is
+far lower than the "target" signatures per slot.
+
+Instead of using the number of signatures to rate govern, the cost model will
+feed back information based on the batch/queue load it is seeing.  The fees will
+sit at a target rate and only increase if the load goes above a specified but to
+be determined threshold.  The governing will be applied across all the fee
+criteria.
+
+### Deterministic fees
+
+Solana's fees are currently deterministic based on a given blockhash.  This
+determinism is a nice feature that simplifies client interactions.  An example
+is when draining an account that is also the payer, the transaction issuer can
+pre-compute the fee and then set the entire remaining balance to be transferred
+out without worrying that the fee will change leaving a very small amount
+remaining in the account.  Another example is for offline signing, the payer
+signer can guarantee what fee that will be charged for the transaction based on
+the nonce's blockhash.
+
+Determinism is achieved in two ways:
+- blockhash queue contains a list of recent (<=~2min) blockhashes and a
+  `lamports_per_signature` value.  The blockhash queue is one of the snapshot's
+  serialized members and thus bank hash depends on it.
+- Nonce accounts used for offline signing contain a `lamports_per_signature`
+  value in its account data
+
+In both cases, when a transaction is assessed a fee, the
+`lamports_per_signature` to use is looked up (either in the queue or in the
+nonce account's data) using the transaction's blockhash.
+
+This currently comes with the following challenges:
+- Exposing the `FeeCalculator` object to the clients (holds the
+  `lamports_per_signature`) makes it hard to evolve the fee criteria due to
+  backward-compatibility.  This issue is being solved by deprecating the
+  `FeeCalculator` object and instead the new apis take a message and return a
+  fee.
+- Blockhash queue entries contain the fee criteria specifics and are part of the
+  bankhash so evolving the fees over time involves more work/risk
+- Nonce accounts store the fee criteria directly in their account data so
+  evolving the fees over time requires changes to nonce account data and data
+  size.
+
+Two solutions to the latter two challenges
+- Get rid of the concept of deterministic fees.  Clients ask via RPC to
+  calculate the current fee estimate and the actual fee is assessed when the
+  transaction is processed.  Fee changes will be governed and change slowly
+  based on network load so the fee differences will be small within the 2min
+  window.  Nonce accounts no longer store the fee criteria but instead a fee
+  cap.  If the assessed fee at the time of processing exceeds the cap then the
```

Comment:
Yeah, for offline advances of the nonce account this would be true, the nonce used and being advanced may have a cap that is insufficient to cover the current fee value.  But the nonce could be advanced with a non-offline-signed transaction (doesn't use the nonce cap).